src/olympia/github/utils.py

```diff
@@ -0,0 +1,149 @@
+            log.info('Body: {}'.format(data))
+            res = requests.post(
+                url,
+                json=data,
+                auth=(settings.GITHUB_API_USER, settings.GITHUB_API_TOKEN))
+            log.info('Response: {}'.format(res.content))
+            res.raise_for_status()
+
+    def pending(self):
+        self.post(self.data['status_url'], data={'state': 'pending'})
+
+    def success(self, url):
+        self.post(self.data['status_url'], data={
+            'state': 'success',
+            'target_url': url
+        })
+
+    def error(self, url):
+        self.post(self.data['status_url'], data={
+            'state': 'error',
+            # Not localising because we aren't sure what locale to localise to.
+            # I would like to pass a longer string here that shows more details
+            # however, we are limited to "A short description of the status."
+            # Which means all the fancy things I wanted to do got truncated.
+            'description': 'This add-on did not validate.',
+            'target_url': url
+        })
+
+    def failure(self):
+        data = {
+            'state': 'failure',
+            # Not localising because we aren't sure what locale to localise to.
+            'description': 'The validator failed to run correctly.'
+        }
+        self.post(self.data['status_url'], data=data)
+
+
+class GithubRequest(forms.Form):
+    status_url = forms.URLField(required=False)
+    zip_url = forms.URLField(required=False)
+    sha = forms.CharField(required=False)
+
+    @property
+    def repo(self):
+        return self.data['pull_request']['head']['repo']
+
+    @property
+    def sha(self):
+        return self.data['pull_request']['head']['sha']
+
+    def get_status(self):
+        return self.repo['statuses_url'].replace('{sha}', self.sha)
+
+    def get_zip(self):
+        return (
+            self.repo['archive_url']
+            .replace('{archive_format}', 'zipball')
+            .replace('{/ref}', '/' + self.sha))
+
+    def validate_url(self, url):
+        if not url.startswith('https://api.github.com/'):
+            raise forms.ValidationError('Invalid URL: {}'.format(url))
+        return url
+
+    def clean(self):
+        fields = (
+            ('status_url', self.get_status),
+            ('zip_url', self.get_zip),
+        )
+        for url, method in fields:
+            try:
+                self.cleaned_data[url] = self.validate_url(method())
+            except:
+                log.error('Invalid data in processing JSON')
+                raise forms.ValidationError('Invalid data')
+
+        self.cleaned_data['sha'] = self.data['pull_request']['head']['sha']
+        self.cleaned_data['type'] = 'github'
+        return self.cleaned_data
+
+
+def rezip_file(response, pk):
+    # An .xpi does not have a directory inside the zip, yet zips from github
+    # do, so we'll need to rezip the file before passing it through to the
+    # validator.
```

Comment:
Help me a bit understanding this, where is the zip coming from GitHub? Are we talking about the release-zips?